cpmpy/expressions/globalconstraints.py

```diff
@@ -496,3 +532,114 @@ def deepcopy(self, memodict={}):
+# Note that count is a numeric global constraint like min/max/element.
+# see commented use in gcc
+# for the decomposition to work, see `decompose_comparison()` of element.
+# then we would have `def atleast(n, vars, val): return Count(vars, val) >= n`
+#is in?
+def alldifferent_except_0(args):
+    return [ ((var1 != 0) & (var2 != 0)).implies(var1 != var2) for var1, var2 in all_pairs(args)]
+
+def count(a,val,c):
+    """
+    Ensure that the number of occurrences of val in a is exactly c.
+    """
+    return [sum(a == val) == c]
+
+def global_cardinality_count(a,gcc):
+    """
+    Global cardinality count: Collect the number of occurrences of each value 0..a.ub in gcc.
+    The array gcc must have elements 0..ub (so of size ub+1).
+    """
+    ub = max([v.ub for v in a])
+    assert (len(gcc) == ub+1), f"GCC: length of gcc variables {len(gcc)} must be ub+1 {ub+1}"
+    return [count(a,i,v) for i, v in enumerate(gcc)]
+    #return [Count(a,i) == v for i, v in enumerate(gcc)]
+
+
+class GlobalCardinalityCount(GlobalConstraint):
+    """
+        GlobalCardinalityCount(a,gcc): Collect the number of occurrences of each value 0..a.ub in gcc.
+    The array gcc must have elements 0..ub (so of size ub+1).
+        """
+
+    def __init__(self, a, gcc):
+        super().__init__("gcc", [a,gcc])
+
+    def decompose(self):
+        a, gcc = self.args
+        ub = max([v.ub for v in a])
+        assert (len(gcc) == ub + 1), f"GCC: length of gcc variables {len(gcc)} must be ub+1 {ub + 1}"
+        return [Count(a, i) == v for i, v in enumerate(gcc)]
+
+
+    def value(self):
+        a, gcc = self.args
+        gval = [argval(y) for y in gcc]
+        return all([gval[i] == Count(a,i).value() for i in range(len(gcc))])
+
+    def deepcopy(self, memodict={}):
+        """
+            Return a deep copy of the AllDifferentExceptO global constraint
+            :param: memodict: dictionary with already copied objects, similar to copy.deepcopy()
+        """
+        copied_args = self._deepcopy_args(memodict)
+        return GlobalCardinalityCount(*copied_args)
+
+class Count(GlobalConstraint):
+    """
+    The Count (numerical) global constraint represents the number of occurrences of val in arr
+    """
+
+    def __init__(self,arr,val):
+        super().__init__("count", [arr,val], is_bool=False)
+
+    def value(self):
+        arr, val = self.args
+        val = argval(val)
+        argvals = [argval(a) for a in arr]
+        return sum([argv == val for argv in argvals])
+
+    def deepcopy(self, memodict={}):
+        arr, val = self._deepcopy_args(memodict)
+        return Count(arr, val)
+
+    def decompose_comparison(self, cmp_op, cmp_rhs):
+        """
+        Count(arr,val) can only be decomposed if it's part of a comparison
+        """
+        from .python_builtins import any
+
+        arr, val = self.args
+        return [any((eval_comparison(cmp_op, j, cmp_rhs) & (Operator('sum',arr == val) == j) for j in range(len(arr) + 1)))]
```

Comment:
Would `eval_comparison(cmp_op, Operator('sum',arr==val), cmp_rhs)` work? The `any` seems unnecessary.